### PR TITLE
Bandwidth estimation

### PR DESCRIPTION
**Pull request by djpetti**
Lots of stuff here...

This adds a more intelligent routing algorithm that makes routing choices based on observed bandwidths. It's quite rudimentary now, but I plan to improve it in the future.

There are also a couple of integration tests for this.

---

This change is [Reviewable](https://reviewable.io/reviews/autoroute/node/76)

### CODE REVIEW

File: internal/bandwidth_test.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	weight_total := float64(0)
+	for _, weight := range estimator.GetWeights(nodes) {
+		if weight-0.5 > 0.001 {
+			t.Fatalf("Expected weight of 0.5, got weight of %f\n", weight)
+		}
+		weight_total += weight
+	}
+
+	// The weights should also approximately add to 1.
+	if weight_total-1.0 > 0.001 {
```
Comment (c00w) [inline internal/bandwidth_test.go]: you should check the absolute value of the difference.

File: integration_tests/autoroute_binary.go [hunk 2]
```diff
@@ -70,11 +70,22 @@ func ProduceCommandLine(b BinaryOptions) []string {
 
 // Produces a AutoRoute Binary which can be run. Must call Start in order to
 // make it start running.
-func NewNodeBinary(b BinaryOptions) AutoRouteBinary {
+// Args:
+//  b: Options for the binary.
+//  race: Whether to use the binary with race condition checking enabled.
+// Returns:
+//  The new autoroute binary.
+func NewNodeBinary(b BinaryOptions, race bool) AutoRouteBinary {
```
Comment (c00w) [inline integration_tests/autoroute_binary.go]: Can you just pass race as part of the binary options?

File: integration_tests/autoroute_binary.go [hunk 3]
```diff
@@ -70,11 +70,22 @@ func ProduceCommandLine(b BinaryOptions) []string {
...
 	port := GetUnusedPort()
 	args := ProduceCommandLine(b)
 	args = append(args, "--status=[::1]:"+fmt.Sprint(port))
-	binary := NewWrappedBinary(GetAutoRoutePath(), args...)
+
+	path := GetAutoRoutePath()
+	if !race {
+		path = GetAutoRoutePathNoRace()
```
Comment (c00w) [inline integration_tests/autoroute_binary.go]: Seems like it'd be simpler to just add compilation options to GetAutoRoutePath where you can specify things like race, rather than having 2 calls.

File: internal/routing.go [hunk 4]
```diff
@@ -90,14 +98,46 @@ func (r *routingHandler) sendPacket(p types.Packet, src types.NodeAddress) error
 		go r.notifyDecision(p, src, r.pk.Hash())
 		return nil
 	}
-	next, err := r.reachability.FindNextHop(p.Destination(), src)
+	possible_next, err := r.reachability.FindNextHop(p.Destination(), src)
```
Comment (c00w) [inline internal/routing.go]: It seems like all of this should be encapsulated within the bandwidth module. Maybe a better way to organize this is

reachability module, which provides information about possible paths based on topology.
bandwidth module, which provides information about current bandwidth + historical bandwidth to various nodes.
Some bandwidth based routing module, which consumes both data sources and does what you're doing here about picking the best.
A super simple routing module which just selects the first next hop from the routing module.

That way we can compare both of them and actually make graphs about which one is better.

File: internal/routing.go [hunk 5]
```diff
@@ -90,14 +98,46 @@ func (r *routingHandler) sendPacket(p types.Packet, src types.NodeAddress) error
...
+	r.bandwidth_mutex.RLock()
+	weights := r.bandwidths.GetWeights(possible_next)
+	r.bandwidth_mutex.RUnlock()
+	// Choose a random destination given our weights.
+	choice := rand.Float64()
+	// We pick the last one because if it should be the last one, floating-point
+	// weirdness could cause us not to pick it.
+	next := possible_next[len(possible_next)-1]
+	total := 0.0
+	for i, weight := range weights {
```
Comment (c00w) [inline internal/routing.go]: unit tests for this?

**Comment by c00w** [comment]
Took a look, I'd like to refactor how this is done so that we can swap the routing decision module relatively easy (especially given that I'd like to swap in a ML module at some point).

Other than that mostly minor nits.

**Comment by c00w** [comment]
Also this is purely a suggestion, but it might make sense to export your variables for the bandwidth estimation so debugging is easier.

Also I should write a status page.

**Comment by djpetti** [comment]
Yeah, I feel weird about exporting that. I can always do it if it becomes a problem. Also, I think I've implemented your idea to make the routing algorithm more modular.

---

Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

_[integration_tests/autoroute_binary.go, line 78 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1N-r1-78:-KEYeKaUZ7-4t0m7IAHh:1246414952) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/integration_tests/autoroute_binary.go#L78)):_
Oh true, I guess it technically is a binary option...

---

_[integration_tests/autoroute_binary.go, line 85 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1N-r1-85:-KEYgNp_xZax-MexSf0Y:-499146009) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/integration_tests/autoroute_binary.go#L85)):_
Done.

---

_[internal/bandwidth_test.go, line 38 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1d-r1-38:-KEYhpeipPnuVCb4WHfM:422879057) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/bandwidth_test.go#L38)):_
Good catch.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEYiiRLMAIzX61CYvhP:1912433808) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
Okay, I elect that we make a whole new "routing" package, and stick all these different routers in there. Otherwise, we'll have at least three files with very similar names. I still personally have trouble remembering the difference between "router.go" and "routing.go"...

---

_[internal/routing.go, line 117 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-117:-KEdEj64EK2rePr7k8TI:-499146009) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L117)):_
Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76#-:-KEdEln7qVI7VfJytUrw:1721867237)_

**Comment by djpetti** [comment]
Review status: 0 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

_[a discussion](https://reviewable.io:443/reviews/autoroute/node/76#-KEdF6930Q2VIFckslig:-KEdF6930Q2VIFckslih:-2012254112) (no related file):_
Also, I've squashed some of the related commits together, so that should make things a little clearer.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by c00w** [comment]
Reviewed 18 of 18 files at r1.
Review status: 7 of 23 files reviewed at latest revision, 2 unresolved discussions.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEe0j4aqILHr7VS8u87:-239038376) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
So we really should just keep the routing module and swap the algorithm inside of it. There isn't any need to conflate the algorithm which makes the decision with the complicated logic about handling packets since that should all be common between all algorithms.

Basically, don't have bandwidth_routing contain a basic_routing, just define an interface with FindNextHop(packet) and have routing call it, and have two implementations, one with bandwidth_algorithm and one with basic_algorithm

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by dcrck** [comment]
Review status: 7 of 23 files reviewed at latest revision, 2 unresolved discussions.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEe1y1woWIcRQ5lSQUU:-22190048) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
Call the interface "RoutingAlgorithm" or something like that. I agree, an interface would make the process of adding new algorithms less complex.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by djpetti** [comment]
Review status: 7 of 23 files reviewed at latest revision, 2 unresolved discussions.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEhu2G77_ynkatNE7kt:-275504970) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
Logistically, that is a little tricky, because for the bandwidth-based algorithm to work, it has to make some special internal calls to a bandwithEstimator when it's actually sending the packet. I'll do my best, though.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by c00w** [comment]
:lgtm:

---

Reviewed 4 of 11 files at r2.
Review status: 9 of 24 files reviewed at latest revision, 2 unresolved discussions.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEjxpujqnQxt8LzCuIB:-743755233) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
So if you split bandwidth _tracking_ from the bandwidth algorithm this should be doable. We actually have an interface for streaming all packet sending decisions out (mostly used for tracking who we owe money to but should be extendable for bandwidth). see the notifyDecision stuff in routing.go. And it makes sense to keep bandwidth around not just for this algo but also for use by other algorithms.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76#-:-KEjxazfvCB8EHVFA7-H:1417088721)_

**Comment by c00w** [comment]
Reviewed 6 of 11 files at r2, 1 of 6 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by c00w** [comment]
_[a discussion](https://reviewable.io:443/reviews/autoroute/node/76#-KEdF6930Q2VIFckslig:-KEjyrauEGzGL0GpjfDO:160371012) (no related file):_
Ack

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_

**Comment by djpetti** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[internal/routing.go, line 101 [r1]](https://reviewable.io:443/reviews/autoroute/node/76#-KETvZ0NAa21-VDUbu1_-r1-101:-KEkv_oJkeP2SbncH8T4:-660232414) ([raw file](https://github.com/autoroute/node/blob/47e4af4bb7337401393f54050d7683957d78f0b8/internal/routing.go#L101)):_
Yeah, the two are already split. (bandwidthEstimator vs. bandwidthRouting) It's hard, though, because we really need two different notifications: One right before we send the packet, and one right after. I think the way I have it now is a reasonable solution.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/76)_
